File: Problems/2131_Longest_Palindrome_by_Concatenating_Two_Letter_Words.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <algorithm>

using namespace std;

class Solution {
public:
    int longestPalindrome(vector<string>& words) {
        unordered_map<string, int> hm;
        int length = 0;
        for(int i=0; i<words.size(); i++){
            string word = words[i];
            string target = word;
            reverse(target.begin(), target.end());
            if(hm.count(target)&&hm[target]>0){
                hm[target]--;
                length += 4;
            }else{
                hm[word]++;
            }
        }
        for(auto it: hm){
            if(it.second>0 && it.first[0]==it.first[1]){
                length += 2;
                break;
            }
        }
        return length;
    }
};

class Solution {
public:
    int longestPalindrome(vector<string>& words) {
        unordered_map<string, int> hm;
        for (string word : words) {
            ++hm[word];
        }
        int total = 0;
        int same = false;
        for (auto iter : hm) {
            string word = iter.first;
            int cnt1 = iter.second;
            string rev = word;
            reverse(rev.begin(), rev.end());
            if (word == rev) {
                total += (cnt1/2) * 4;
                if (cnt1 % 2)
                    same = true;
            }else {
                if (!hm.count(rev)) {
                    continue;	
                }
                int cnt2 = hm[rev];
                int used = min(cnt1, cnt2);
                total += used * 4;
                hm[word] -= used;
                hm[rev]  -= used;
            }
        }
        return same ? total + 2 : total;
    }
};

File: Problems/2131_Longest_Palindrome_by_Concatenating_Two_Letter_Words.py
from collections import defaultdict
from typing import List

class Solution:
    def longestPalindrome(self, words: List[str]) -> int:
        hm = defaultdict(int)
        for word in words:
            hm[word] += 1
        
        total = 0        
        same = False
        for word, cnt1 in hm.items():
            if word[0]==word[1]:
                total += (cnt1//2) * 4
                if cnt1 % 2:
                    same = True
            else:
                if word[::-1] not in hm:
                    continue
                cnt2 = hm[word[::-1]]
                used =  min(cnt1, cnt2)
                hm[word] -= used
                hm[word[::-1]] -= used
                total += used* 4
        return total + 2 if same else total


if __name__ == '__main__':
    sol = Solution()
    words = ["lc","cl","gg"]
    print(sol.longestPalindrome(words))